docs/source/expr-system-internals.rst

```diff
@@ -0,0 +1,206 @@
+Query planning with Expression system
+=====================================
+
+.. note::
+
+    This document is intended for Dask developers and contributors. It is not
+    intended for end-users.
+    For a high level user guide, see :doc:`/dataframe-optimizer`.
+
+.. currentmodule:: dask.dataframe
+
+The expression system was originally developed for Dask DataFrames, as implemented in the `dask-expr <https://github.com/dask/dask-expr>`_ project. Early prototypes experimented with `matchpy <https://github.com/HPAC/matchpy>`_, but it was soon replaced with a custom-built system offering simpler and more transparent optimization steps.
+
+Expr objects
+------------
+
+The expression system is built around the `Expr` class. This class is used to represent a computation that can be performed on a Dask DataFrame. The `Expr` class is designed to be subclassed, and each subclass represents a specific type of computation. For example, there are subclasses for arithmetic operations, logical operations, and so on.
+
+Construction
+^^^^^^^^^^^^
+
+The expression system centers around the Expr class, which represents a computation on a Dask DataFrame. This class is designed for subclassing; each subclass corresponds to a specific computation type (e.g., arithmetic, logical operations, filtering, joins).
+
+Notably, custom initializers (``__init__``) are disallowed * both in the base class and its subclasses. This design decision reflects concerns around performance, as expression objects may be created and recreated frequently, and custom logic in constructors could introduce unnecessary overhead.
+
+Instead, expression classes use a dataclass-like interface defined by two attributes:
+
+* ``_parameters``: List of parameter names
+
+* ``_defaults``: Dictionary of default values for optional parameters
```

Comment:
This stuff seems good, and probably more important than the note about custom initializers.  I recommend leading with it, and probably with other structural parts of the expression system (children, expected methods, etc.) before we get into caveats.

Reply:
Let me ask a broader question then, if someone wanted to write an Expression powered version of some other API (maybe Bag or something) what would they read in order to learn how to do that?  Is there a doc already?

Reply:
I don't know exactly how that works so it is difficult to write down documentation for that

Reply:
You don't know how the expression system works, or how bag works?

I think you know how the expression system works.  My request is that you write down the information you would want someone to know if they were to start trying to build a new API on top of it.  Why might they want to do that?  What are the key parts of the interface they should be aware of?  etc..

Reply:
I updated the old examples about collections and added one using Expressions

Reply:
which also includes an overdue update regarding the task spec